contrib/src/main/java/macrobase/analysis/summarize/ASAP.java

```diff
@@ -0,0 +1,234 @@
+package macrobase.analysis.summarize;
+
+import macrobase.analysis.pipeline.stream.TimeDatumStream;
+import macrobase.analysis.stats.Autocorrelation;
+import macrobase.analysis.summarize.util.ASAPMetrics;
+import macrobase.analysis.transform.BatchSlidingWindowTransform;
+import macrobase.analysis.transform.aggregate.AggregateConf;
+import macrobase.conf.ConfigurationException;
+import macrobase.conf.MacroBaseConf;
+import macrobase.datamodel.Datum;
+
+import java.util.List;
+
+public class ASAP {
+    public List<Datum> currWindow;
+    public long binSize;
+    public int windowSize = 1;
+    public int pointsChecked = 0;
+
+    /* Do not smooth for time series whose initial kurtosis is above this threshold */
+    private double KURT_THRESH = 5;
+    /* Search for window sizes up to #points / MAX_WINDOW */
+    private int MAX_WINDOW = 10;
+    /* Kurtosis to preserve after smoothing */
+    private double PRESERVATION_THRESH = 1;
+
+    private BatchSlidingWindowTransform swTransform;
+    private MacroBaseConf conf = new MacroBaseConf();
+    private Autocorrelation acf;
+    private ASAPMetrics metrics;
+    private TimeDatumStream stream;
+    private double minObj;
+    private boolean preAggregate;
+
+    public ASAP(TimeDatumStream stream, long windowRange, int resolution, boolean preAggregate) throws Exception {
+        this.stream = stream;
+        this.preAggregate = preAggregate;
+        conf.set(MacroBaseConf.TIME_COLUMN, stream.timeColumn);
+        conf.set(AggregateConf.AGGREGATE_TYPE, AggregateConf.AggregateType.AVG);
+
+        currWindow = stream.drainDuration(windowRange);
+        // Right now ASAP only supports a timestamp dimension and a metrics dimension
+        assert (currWindow.get(0).metrics().getDimension() == 2);
+        binSize = getBinSize(windowRange, resolution);
+        if (preAggregate) {
+            preAggregate();
+        } else {
+            metrics = new ASAPMetrics(currWindow, 1 - stream.timeColumn);
+        }
+
+        acf = new Autocorrelation(currWindow.size() / MAX_WINDOW, 1 - stream.timeColumn);
+        // Only search for time series whose initial Kurtosis is not too large
+        if (metrics.originalKurtosis < KURT_THRESH) { acf.evaluate(currWindow); }
+
+    }
+
+    public void setKurtThresh(double thresh) { KURT_THRESH = thresh; }
+
+    public void setMaxWindow(int maxWindow) { MAX_WINDOW = maxWindow; }
+
+    public void setPreservationThresh(double thresh) { PRESERVATION_THRESH = thresh; }
+
+    /* Get bin size given time period and resolution to visualize.
+    *  Round bin size to the nearest 10min / hour / day. */
+    private long getBinSize(long windowRange, int resolution) {
+        long binSize = windowRange / resolution;
+        long dayInSec = 24 * 3600 * 1000L;
```

Comment:
prefer timeunit? http://stackoverflow.com/a/924221
also, should be static final?